### PR TITLE
SSP-2030_OIDC_module_switch_to_ProcessingChain_for_authproc_support_consent_mod

### PR DESCRIPTION
**Pull request by ioigoume**
#179

### CODE REVIEW

**Comment by ioigoume** [comment]
@cicnavi 
the last commit on the `wip-version-6` branch broke the dependencies injected from the `modules/oidc/src/Services/Container.php` class.

**Comment by cicnavi** [comment]
> @cicnavi the last commit on the `wip-version-6` branch broke the dependencies injected from the `modules/oidc/src/Services/Container.php` class.

What exactly, how should I reproduce this?

**Comment by ioigoume** [comment]
@cicnavi ,
before the commit that changed the routes this is how the authorize flow worked:
- public/authorize.php
- RoutingService initialized the ContainerService and created and injected all the dependencies to all the other services
- Called the controller

After the change the module uses the Symfony routing service. As a result, the old ContainerService is never called and the dependencies are not created/injected in the controller as before. For example the `$requestRuleManager` dependency is null.

In order to reproduce I am running the authorization flow.

**Comment by cicnavi** [comment]
Well yes, we are moving to Symfony routes and container. In v6, I was thinking to have both Symfony routes and "old" public/*.php routes available. The reason for both routes is that some RPs will have "old" configuration fetched from the configuration endpoint, and in this way they will still be able to authenticate using old routes and have time to fetch new configuration with new endpoints. 

So, it's important that the old routes still function if they are used. I thought that I messed something up with them, but if I understand correctly, the problem for you now is that they are not used by default any more. 

So, yes, if you are going to implement new stuff, please use Symfony routes and container... this is how SSP now handles requests, and we are actually pretty late to implement this...

**Comment by cicnavi** [comment]
Oh, and if I can help in any way, please let me now. Feel free to ping me on Slack (join group SimpleSAMLphp if you haven't already).

**Comment by ioigoume** [comment]
@cicnavi 
this is the stack trace reported when I try to access the user info new endpoint

Screenshot from 2024-06-12 11-01-31

**Comment by ioigoume** [comment]
@cicnavi can you please provide some feedback on this PR. I refactored the existing tests so that they run correctly. As soon as we have a consensus about the new code I will extend the tests.

**Comment by cicnavi** [comment]
Will do as soon as I can

File: src/Controller/AuthorizationController.php [hunk 1]
```diff
@@ -56,27 +57,43 @@ public function __construct(
      */
     public function __invoke(ServerRequestInterface $request): ResponseInterface
     {
-        $authorizationRequest = $this->authorizationServer->validateAuthorizationRequest($request);
+        $queryParameters = $request->getQueryParams();
+        if (!isset($queryParameters[ProcessingChain::AUTHPARAM])) {
+            $authorizationRequest = $this->authorizationServer->validateAuthorizationRequest($request);
+            $this->authenticationService->processRequest($request, $authorizationRequest);
```
Comment (pradtke) [inline src/Controller/AuthorizationController.php:63]: My preference would be for `processRequest` to not redirect unless one of the authproc filters issues one. I think you can achieve that by having it return the $state array, not call ProcessingComplete and then put ` $state = $this->authenticationService->loadState($queryParameters);` into the else.

My reason for preferring this is that the vast majority of our logins do not need to redirect the user, and adding the extra redirect makes the system feel slower.

File: src/Services/AuthenticationService.php [hunk 2]
```diff
@@ -111,23 +134,57 @@ public function getAuthenticateUser(
...
+     * @param   array  $state
+     *
+     * @return AuthorizationRequest
+     * @throws Exception
+     */
+
+    public function getAuthorizationRequestFromState(array $state): AuthorizationRequest
+    {
+        if (!($state['authorizationRequest'] instanceof AuthorizationRequest)) {
+            throw new Exception("Authorization Request is not valid: " . print_r($state, true));
```
Comment (pradtke) [inline src/Services/AuthenticationService.php:169]: I don't think the `$state` value should appear in the exception. I think there is often data in the `$state` array that we don't want to appear in exception messages.  For example we (Cirrus) don't want a user's personal information to appear in our logs, and that PII is in the `$state` array and will be part of the exception message.

File: src/Services/AuthenticationService.php [hunk 3]
```diff
@@ -173,7 +232,31 @@ public function getSessionId(): ?string
...
+
+    public function authenticate(
+        ServerRequestInterface $request,
+        array $loginParams = [],
+    ): void {
+        $oidcClient = $this->getClientFromRequest($request);
+        $authSimple = $this->authSimpleFactory->build($oidcClient);
+
+        $this->sessionService->setIsCookieBasedAuthn(false);
+        $this->sessionService->setIsAuthnPerformedInPreviousRequest(true);
```
Comment (pradtke) [inline src/Services/AuthenticationService.php:253]: @cicnavi I'm not familiar with the purpose of `setIsAuthnPerformedInPreviousRequest`, so I'll let you leave feedback if this is correct.  It looks correct to me solely based on splitting `getAuthenticatedUser` apart into more specific methods

File: src/Services/AuthenticationService.php [hunk 4]
```diff
@@ -190,4 +273,61 @@ protected function addRelyingPartyAssociation(ClientEntityInterface $oidcClient,
...
+            // ProcessChain needs to know the list of authproc filters we defined in module_oidc configuration
+            'authproc' => $this->moduleConfig->getAuthProcFilters(),
+        ];
+        $spMetadata = [
+            'entityid' => $state['Destination']['entityid'] ?? '',
+        ];
+        $pc = new ProcessingChain(
+            $idpMetadata,
+            $spMetadata,
+            explode('.', (string)$this->moduleConfig::OPTION_AUTH_PROCESSING_FILTERS)[1],
```
Comment (pradtke) [inline src/Services/AuthenticationService.php:320]: I think this is more readable if we remove the whole `explode` bit and just use the string `oidc`

File: src/Services/AuthenticationService.php [hunk 5]
```diff
@@ -190,4 +273,61 @@ protected function addRelyingPartyAssociation(ClientEntityInterface $oidcClient,
...
+
+        $state['ReturnURL'] = $this->moduleConfig->getModuleUrl(RoutesEnum::OpenIdAuthorization->value);
+        $state['Destination'] = $spMetadata;
+        $state['Source'] = $idpMetadata;
+
+        $pc->processState($state);
+
+        // If no filter(s) is available in the configuration, it will enforce a redirect
+        // to the authentication endpoint.
+        ProcessingChain::resumeProcessing($state);
```
Comment (pradtke) [inline src/Services/AuthenticationService.php:331]: Let's remove the resumeProcessing so we don't need to do the extra redirect.

File: src/Utils/Checker/Rules/MaxAgeRule.php [hunk 6]
```diff
@@ -75,7 +75,7 @@ public function checkRule(
             $loginParams = [];
             $loginParams['ReturnTo'] = (new HTTP())->addURLParameters((new HTTP())->getSelfURLNoQuery(), $queryParams);
 
-            $this->authenticationService->getAuthenticateUser($request, $loginParams, true);
+            $this->authenticationService->authenticate($request, $loginParams);
```
Comment (pradtke) [inline src/Utils/Checker/Rules/MaxAgeRule.php:78]: I like this change. I think it makes it more obvious that we are just making the user authenticate rather than wanting to get the user object.

File: tests/src/Services/AuthenticationServiceTest.php [hunk 7]
```diff
@@ -199,45 +174,63 @@ public function testItReturnsAnUser(): void
...
+    /**
+     * @return void
+     * @throws \JsonException
+     * @throws \SimpleSAML\Error\BadRequest
+     * @throws \SimpleSAML\Error\NotFound
+     * @throws \SimpleSAML\Module\oidc\Server\Exceptions\OidcServerException
+     */
+    public function testItAuthenticates(): void
+    {
+        $authenticationServiceMock = $this->getMockBuilder(AuthenticationService::class)
```
Comment (pradtke) [inline tests/src/Services/AuthenticationServiceTest.php:214]: I don't follow why you are mocking the class under tests.  I think you can use use the an instance of AuthenticationService with mocked dependencies, which is what `$this->prepareMockedInstance()` should give you.

**Review by pradtke — CHANGES_REQUESTED** [review]
I think it looks pretty good. I'd like to remove the redirect that happens from calling `ProcessingChain::resumeProcessing($state);`

I think for tests, I think you need one for `AuthorizationControllerTest` that invokes the code path that calls `authenticationService->processRequest`  and a test in `AuthenticationServiceTest` to test `processRequest`

**Comment by cicnavi** [comment]
There are some conformance test issues, but are not visible here since no action runs were started. @tvdijen what is the policy on running actions in PRs?

**Comment by cicnavi** [comment]
@ioigoume You can also try to run conformance tests locally as per https://github.com/simplesamlphp/simplesamlphp-module-oidc/blob/wip-version-6/CONFORMANCE_TEST.md

File: src/Services/AuthenticationService.php (hunk 2 above)
Comment (ioigoume) [reply src/Services/AuthenticationService.php:169]: done

File: src/Services/AuthenticationService.php (hunk 4 above)
Comment (ioigoume) [reply src/Services/AuthenticationService.php:320]: done

File: src/Controller/AuthorizationController.php (hunk 1 above)
Comment (ioigoume) [reply src/Controller/AuthorizationController.php]: done

File: src/Services/AuthenticationService.php (hunk 5 above)
Comment (ioigoume) [reply src/Services/AuthenticationService.php]: done

**Comment by tvdijen** [comment]
> [..] @tvdijen what is the policy on running actions in PRs?

We do this on any other repository, but this repo's test-suite has diverged over the years and I stopped keeping it in sync

File: src/Entities/AccessTokenEntity.php [hunk 8]
```diff
@@ -157,13 +158,20 @@ public function setRequestedClaims(array $requestedClaims): void
      */
     public function getState(): array
     {
+        $database = Database::getInstance();
```
Comment (pradtke) [inline src/Entities/AccessTokenEntity.php:161]: I don't understand why this is changing.  I think the PR should be focussed on things related to the processing chain.  I think if there are issues with the database unrelated to the processing chain then that should go in a different PR so the conversation can be more focussed.

**Review by pradtke — CHANGES_REQUESTED** [review]
I think we also still need tests.

**Comment by cicnavi** [comment]
> > [..] @tvdijen what is the policy on running actions in PRs?
> 
> We do this on any other repository, but this repo's test-suite has diverged over the years and I stopped keeping it in sync

I don't follow. When I do a pull request, actions do run. Why they didn't run for ioigoumes PR?

File: src/Entities/AccessTokenEntity.php (hunk 8 above)
Comment (ioigoume) [reply src/Entities/AccessTokenEntity.php:161]: I am getting errors when i run the conformance tests, as a result i included the fixes in this PR. I will move them in a separate PR.

**Comment by tvdijen** [comment]
> I don't follow. When I do a pull request, actions do run. Why they didn't run for ioigoumes PR?

Ah! Because they only run against `master` or branches called `release-*`..

https://github.com/simplesamlphp/simplesamlphp-module-oidc/blob/master/.github/workflows/test.yaml#L7

We could add `bugfix/* ` and `feature/*` but it will require some discipline to keep the naming-convention

File: src/Entities/AccessTokenEntity.php (hunk 8 above)
Comment (ioigoume) [reply src/Entities/AccessTokenEntity.php]: moved to a different issue

**Comment by cicnavi** [comment]
> > I don't follow. When I do a pull request, actions do run. Why they didn't run for ioigoumes PR?
> 
> Ah! Because they only run against `master` or branches called `release-*`..
> 
> https://github.com/simplesamlphp/simplesamlphp-module-oidc/blob/master/.github/workflows/test.yaml#L7
> 
> We could add `bugfix/* ` and `feature/*` but it will require some discipline to keep the naming-convention

Oh, right. But it wouldn't make a difference since he is having those branches in its own fork. I would rather run them on any PR

> > I don't follow. When I do a pull request, actions do run. Why they didn't run for ioigoumes PR?
> 
> Ah! Because they only run against `master` or branches called `release-*`..
> 
> https://github.com/simplesamlphp/simplesamlphp-module-oidc/blob/master/.github/workflows/test.yaml#L7
> 
> We could add `bugfix/* ` and `feature/*` but it will require some discipline to keep the naming-convention

But again, that wouldn't run them here since he has them in his own fork I guess... Would it be ok for you that we run those on all PRs? Something like:
```
on:
  pull_request:
    types: [opened, synchronize, reopened, closed]
```

**Comment by ioigoume** [comment]
> @ioigoume You can also try to run conformance tests locally as per https://github.com/simplesamlphp/simplesamlphp-module-oidc/blob/wip-version-6/CONFORMANCE_TEST.md

@cicnavi thank you for the hint. I managed to run the tests manually directly from the `https://staging.certification.openid.net` site

File: src/Services/AuthenticationService.php [hunk 9]
```diff
@@ -79,7 +79,7 @@ public function __construct(
     public function processRequest(
         ServerRequestInterface $request,
         OAuth2AuthorizationRequest $authorizationRequest,
-    ): void {
+    ): null|array {
```
[the remark this answers is not in the document]
Comment (ioigoume) [reply src/Services/AuthenticationService.php]: that was wrong. I fixed it after I ran the psalm tests.

**Comment by tvdijen** [comment]
> But again, that wouldn't run them here since he has them in his own fork I guess... [..]

No, no, it is because of the name of the target branch!

> Would it be ok for you that we run those on all PRs

Like I said, the testsuite here is much different than the rest of the repositories, and I consider this repo yours... do as you please!

File: tests/src/Services/AuthenticationServiceTest.php (hunk 7 above)
Comment (ioigoume) [reply tests/src/Services/AuthenticationServiceTest.php:214]: the `$this->prepareMockedInstance()` returns the actual object. What I needed for this test was a mocked object so that  I could use the `expects` function

**Review by pradtke — CHANGES_REQUESTED** [review]
There is the line about $authState being a string that I don't understand.

I also don't see a test for AuthorizationControllerTest that results in `$this->authenticationService->processRequest` being called.

Thanks

File: src/Services/AuthenticationService.php [hunk 10]
```diff
@@ -154,26 +227,73 @@
 
         $state[State::RESTART] = $request->getUri()->__toString();
 
+        // Data required after we get back from a ProcessingChain redirect
+        $state['authorizationRequest'] = $authorizationRequest;
```
Comment (cicnavi) [inline src/Services/AuthenticationService.php:231]: Here League\OAuth2\Server\RequestTypes\AuthorizationRequest as OAuth2AuthorizationRequest is set, while in getAuthorizationRequestFromState() there is a check if the instance is of League\OAuth2\Server\RequestTypes\AuthorizationRequest. I think this may break Server/Grants/OAuth2ImplicitGrant.php, which only uses  OAuth2AuthorizationRequest (haven't tested it yet).

File: src/Controller/AuthorizationController.php [hunk 11]
```diff
@@ -56,27 +57,46 @@
...
+        $user = $this->authenticationService->getAuthenticateUser($state);
 
         $authorizationRequest->setUser($user);
         $authorizationRequest->setAuthorizationApproved(true);
 
-        if ($authorizationRequest instanceof AuthorizationRequest) {
-            $authorizationRequest->setIsCookieBasedAuthn($this->authenticationService->isCookieBasedAuthn());
-            $authorizationRequest->setAuthSourceId($this->authenticationService->getAuthSourceId());
-            $authorizationRequest->setSessionId($this->authenticationService->getSessionId());
+        $authorizationRequest->setIsCookieBasedAuthn($this->authenticationService->isCookieBasedAuthn());
```
Comment (cicnavi) [inline src/Controller/AuthorizationController.php:77]: Note that this method does not exist in OAuth2AuthorizationRequest (League\OAuth2\Server\RequestTypes\AuthorizationRequest). Previously there was a check if the instance is of expected type. The whole thing "stinks", but that's until we introduce proper OAuth2 package briges... :/

File: src/Services/AuthenticationService.php [hunk 12]
```diff
@@ -190,4 +310,64 @@
...
+    {
+        $idpMetadata = [
+            'entityid' => $state['Source']['entityid'] ?? '',
+            // ProcessChain needs to know the list of authproc filters we defined in module_oidc configuration
+            'authproc' => $this->moduleConfig->getAuthProcFilters(),
+        ];
+        $spMetadata = [
+            'entityid' => $state['Destination']['entityid'] ?? '',
+        ];
+        $pc = new ProcessingChain(
```
[the remark this answers is not in the document]
Comment (ioigoume) [reply src/Services/AuthenticationService.php:361]: I will do the implementation.

File: src/Controller/AuthorizationController.php (hunk 11 above)
Comment (ioigoume) [reply src/Controller/AuthorizationController.php:77]: The check is no longer needed since we do it inside the AuthorizationController::getAuthorizationRequestFromState.

File: src/Services/AuthenticationService.php (hunk 12 above)
[the remark this answers is not in the document]
Comment (ioigoume) [reply src/Services/AuthenticationService.php]: Done

File: src/Services/AuthenticationService.php [hunk 13]
```diff
@@ -41,10 +41,20 @@ class AuthenticationService
 {
     use GetClientFromRequestTrait;
 
+    /**
+     * @var \SimpleSAML\Auth\State|string
+     * @psalm-var \SimpleSAML\Auth\State|class-string
+     */
+    protected string|State $authState = \SimpleSAML\Auth\State::class;
```
[the remark this answers is not in the document]
Comment (ioigoume) [reply src/Services/AuthenticationService.php]: done

File: src/Controller/AuthorizationController.php (hunk 11 above)
Comment (ioigoume) [reply src/Controller/AuthorizationController.php:77]: I fixed it. Even though i was checking the instance of the authorizationRequest, i was not taking into account the 
```php
        return $this->authorizationServer->completeAuthorizationRequest(
            $authorizationRequest,
            $this->psrHttpBridge->getResponseFactory()->createResponse(),
        );
```
code

File: src/Services/AuthenticationService.php (hunk 10 above)
Comment (ioigoume) [reply src/Services/AuthenticationService.php:230]: i pushed a fixed and now the `AuthenticationService::getAuthorizationRequestFromState`  returns the correct object instance type.

**Review by pradtke — APPROVED** [review]

**Comment by pradtke** [comment]
I did some more local testing with the authprocs and things worked as I expected.
I made a minor commit to add some info on running a docker image with preprodwarning authproc enabled to test redirects, and I enabled one basic authproc for the conformance tests (just to confirm that invoking authproc doesn't make anything explode, there is no validation of what the authproc does in the conformance test)

@cicnavi I'm ready to merge this. If you agree then I can do a squash and merge.  Thanks

**Comment by cicnavi** [comment]
I haven't looked at it yet, but if you say it's OK, then go ahead. Please
consider noting it in the upgrade log. For example, will this change now
also run any "global" authproc from config.php? If so, someone could have
SAML only authproc set globally which will then break because it won't have
expected state.

**Comment by ioigoume** [comment]
@cicnavi 
the module will only run the authproc filters configured in the `module_oidc.php` file.

```php
        $idpMetadata = [
            'entityid' => $state['Source']['entityid'] ?? '',
            // ProcessChain needs to know the list of authproc filters we defined in module_oidc configuration
            'authproc' => $this->moduleConfig->getAuthProcFilters(),
        ];
```

**Comment by pradtke** [comment]
Good call out on the documentation. I updated it to reflect that we can now support doing redirects in the authproc.
The ProcessinChain does try to load additional authproc filters from the main `config.php` but it will be looking for them under the key `authproc.oidc` rather than `authproc.idp` that saml uses. I made a note of that behavior changes in the upgrade notes incase someone, for some unknown reason, has that key already defined.

**Comment by cicnavi** [comment]
Great, thanks!
